File: domain/use_case/command_use_case.py
from domain.commands.handlers import CommandHandler
from models.config import LLMConfig
from utils.logger import Logger


class CommandUseCase:
    def __init__(
        self,
        command_registry: dict[str, CommandHandler],
        # TODO: Isso daqui vai dar dor de cabeça com o comando de troca de modelo
        llm_config: LLMConfig,
        logger: Logger,
    ):
        self._registry = command_registry
        self._llm_config = llm_config
        self._logger = logger

        self._available_deps = {
            "llm_config": self._llm_config,
            "logger": self._logger,
        }

    def execute(self, command: str) -> CommandHandler:
        """
        Execute the command if it exists in the registry.
        """
        parts = command.strip().split(maxsplit=1)
        # TODO: Por agora o comando não tem argumento :)
        command_name = parts[0]

        handler_func = self._registry.get(command_name)
        if handler_func is None:
            self._logger.warning(
                "Comando desconhecido.",
                context={
                    "command": command_name,
                    "available_commands": list(self._registry.keys()),
                },
            )
            return CommandHandler(
                message=f"Comando desconhecido: {command_name}. Digite /help para ver os comandos disponíveis.",
                should_exit=False,
            )

        try:
            result = handler_func(**self._available_deps)
            self._logger.info(
                "Comando executado com sucesso.",
                context={"command": command_name, "result": result},
            )
            return result
        except Exception as e:
            self._logger.error(
                "Erro ao executar o comando.",
                context={"command": command_name, "error": str(e)},
            )
            return CommandHandler(
                message=f"Erro ao executar o comando {command_name}: {str(e)}",
                should_exit=False,
            )

File: src/domain/use-case/command-use-case.ts
/**
 * Use case for executing CLI commands.
 */

import type { CommandHandler } from '../commands/handlers.js';
import type { CommandResult } from '../models/command.js';
import { createCommandResult } from '../models/command.js';
import type { Logger } from '../../utils/logger.js';
import type { LLMConfig } from '../models/config.js';
import type { WorkingDirectoryManager } from '../pattern/working-directory-manager.js';

export class CommandUseCase {
  constructor(
    private readonly registry: Record<string, CommandHandler>,
    private readonly llmConfig: LLMConfig,
    private readonly logger: Logger,
    private readonly workingDirectoryManager: WorkingDirectoryManager
  ) {}

  execute(command: string): CommandResult {
    const parts = command.trim().split(/\s+(.+)/);
    const commandName = parts[0] ?? '';
    const commandArguments = parts[1] ?? '';

    const handlerFunc = this.registry[commandName];

    if (!handlerFunc) {
      this.logger.warning('Unknown command.', {
        command: commandName,
        arguments: commandArguments,
        availableCommands: Object.keys(this.registry),
      });

      return createCommandResult(
        `Comando desconhecido: ${commandName}. Digite /help para ver os comandos disponíveis.`,
        false
      );
    }

    try {
      const result = handlerFunc({
        commandArguments,
        llmConfig: this.llmConfig,
        logger: this.logger,
        workingDirectoryManager: this.workingDirectoryManager,
      });

      this.logger.info('Command executed successfully.', {
        command: commandName,
        result,
      });

      return result;
    } catch (error) {
      this.logger.error('Error executing command.', {
        command: commandName,
        error: (error as Error).message,
      });

      return createCommandResult(
        `Erro ao executar o comando ${commandName}: ${(error as Error).message}`,
        false
      );
    }
  }
}
